dNR: Implement requestMethods and excludedRequestMethods

elijahsawyers opened the pull request:
#### b2d8209ad8fd896035cd90ed04f9090bb9b54f21
<pre>
dNR: Implement requestMethods and excludedRequestMethods
<a href="https://bugs.webkit.org/show_bug.cgi?id=291595">https://bugs.webkit.org/show_bug.cgi?id=291595</a>
<a href="https://rdar.apple.com/132579559">rdar://132579559</a>

Reviewed by Brian Weinstein.

This patch implements requestMethods and excludedRequestMethods dNR conditions.
To do this, we have to create one WebKit content blocking rule per request
method, and one ignore-previous-rules rule per excluded request method. This is
necessary because the WebKit content blocking trigger for request methods only
accepts a string, not an array of strings. For more information, see:

<a href="https://bugs.webkit.org/show_bug.cgi?id=290962">https://bugs.webkit.org/show_bug.cgi?id=290962</a>

Additionally, this patch adds tests to validate the rule conversion from dNR to
WebKit content blocking.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(WebKit::if):
(isArrayOfRequestMethodsValid):
(-[_WKWebExtensionDeclarativeNetRequestRule ruleInWebKitFormat]):
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:condition:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithRequestMethods)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithExcludedRequestMethods)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithRequestMethodsAndExcludedRequestMethodsAndRequestDomainsAndExcludedRequestDomains)):

Canonical link: <a href="https://commits.webkit.org/293853@main">https://commits.webkit.org/293853@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8dfce90e3eae3b9b91da1671b47037752dd6d730

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100065 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19713 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10006 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105193 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50646 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102106 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20019 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28186 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76174 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33244 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103072 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15289 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90369 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56535 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15103 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8365 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50015 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85016 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8449 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107553 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27178 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19902 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85127 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27541 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86575 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84660 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29331 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7066 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21014 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16289 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27115 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32344 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26926 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30242 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28485 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -702,6 +723,21 @@ static BOOL isArrayOfExcludedDomainsValid(NSArray<NSString *> *excludedDomains)
 
+static BOOL isArrayOfRequestMethodsValid(NSArray<NSString *> *requestMethods)
+{
+    NSSet<NSString *> *validRequestMethods = [NSSet setWithObjects:@"get", @"head", @"options", @"trace", @"put", @"delete", @"post", @"patch", @"connect", nil];
```
Nit: We should probably make this a static set that's defined in a dispatch_once so we don't have to do this work every time `isArrayOfRequestMethodsValid` is called.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -702,6 +723,21 @@ static BOOL isArrayOfExcludedDomainsValid(NSArray<NSString *> *excludedDomains)
+    NSSet<NSString *> *validRequestMethods = [NSSet setWithObjects:@"get", @"head", @"options", @"trace", @"put", @"delete", @"post", @"patch", @"connect", nil];
+
+    if (!requestMethods.count)
+        return NO;
```
And creating `validRequestMethods` could go below this early return.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -364,6 +368,23 @@ - (instancetype)initWithDictionary:(NSDictionary *)ruleDictionary errorString:(N
+    if (NSArray<NSString *> *requestMethods = _condition[declarativeNetRequestRuleConditionRequestMethodsKey]) {
+        if (!isArrayOfRequestMethodsValid(requestMethods)) {
+            if (outErrorString)
+                *outErrorString = [[NSString alloc] initWithFormat:@"Rule with id %ld is invalid. `requestMethods` must be non-empty and can only contain `get`, `head`, `options`, `trace`, `put`, `delete`, `post`, `patch`, and `connect`.", (long)_ruleID];
```
Instead of duplicating the allowed methods, in theory we could create a helper method that returns the set of the allowed methods, and use it in the logging. I'm not sure if that needs to be done, just duplicating the list of them is a bit concerning in case we add any in the future.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -732,29 +768,59 @@ - (void)removeInvalidResourceTypesForKey:(NSString *)ruleConditionKey
+        }
+    } else if (excludedRequestMethods) {
+        for (NSString *excludedRequestMethod in excludedRequestMethods)
+            [convertedRules addObjectsFromArray:[self _convertRulesWithModifiedCondition:createModifiedConditionsForRequestMethod(_condition, excludedRequestMethod) webKitActionType:@"ignore-previous-rules" chromeActionType:chromeActionType]];
```
I'm not sure what to do about it right now, but this method is getting awfully complicated. In future patches we might want to think about how to avoid duplicating this indentation/structure across every case.

elijahsawyers replied:
Yeah, I think we'd eventually just want to support directly compiling dNR rules in byte code rather than having this conversion step.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
```diff
@@ -702,6 +723,41 @@ static BOOL isArrayOfExcludedDomainsValid(NSArray<NSString *> *excludedDomains)
+        validRequestMethodsString = [NSString stringWithFormat:@"%@, and %@", validRequestMethodsString, validRequestMethodsArray.lastObject];
+    });
+    return validRequestMethodsString;
+}
```
Nice. I like how you did this.

elijahsawyers committed:
dNR: Implement requestMethods and excludedRequestMethods
https://bugs.webkit.org/show_bug.cgi?id=291595
rdar://132579559

Reviewed by Brian Weinstein.

This patch implements requestMethods and excludedRequestMethods dNR conditions.
To do this, we have to create one WebKit content blocking rule per request
method, and one ignore-previous-rules rule per excluded request method. This is
necessary because the WebKit content blocking trigger for request methods only
accepts a string, not an array of strings. For more information, see:

https://bugs.webkit.org/show_bug.cgi?id=290962

Additionally, this patch adds tests to validate the rule conversion from dNR to
WebKit content blocking.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(WebKit::if):
(isArrayOfRequestMethodsValid):
(-[_WKWebExtensionDeclarativeNetRequestRule ruleInWebKitFormat]):
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:condition:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithRequestMethods)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithExcludedRequestMethods)):
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, RuleConversionWithRequestMethodsAndExcludedRequestMethodsAndRequestDomainsAndExcludedRequestDomains)):

Canonical link: https://commits.webkit.org/293853@main
